components/public-api/gitpod/experimental/v1/tokens.proto

```diff
@@ -0,0 +1,83 @@
+syntax = "proto3";
+
+package gitpod.experimental.v1;
+
+option go_package = "github.com/gitpod-io/gitpod/components/public-api/go/experimental/v1";
+
+import "google/protobuf/timestamp.proto";
+
+service TokenService {
+
+    // CreateToken creates a new token and returns it.
+    rpc CreateToken(CreateTokenRequest) returns (CreateTokenResponse) {}
+
+    // GetToken returns one token.
+    rpc GetToken(GetTokenRequest) returns (GetTokenResponse) {}
+
+    // ListTokens returns a list of tokens.
+    rpc ListTokens(ListTokensRequest) returns (ListTokensResponse) {}
+
+    // UpdateToken either regenerates a token or updates a token's scopes.
+    rpc UpdateToken(UpdateTokenRequest) returns (UpdateTokenResponse) {}
+
+    // DeleteToken removes a token.
+    rpc DeleteToken(DeleteTokenRequest) returns (DeleteTokenResponse) {}
+}
+
+message CreateTokenRequest {
+    PersonalAccessToken token = 1;
+}
+
+message CreateTokenResponse {
+    PersonalAccessToken token = 1;
+}
+
+message GetTokenRequest {
+    string token_hash = 1;
+    string user_id = 2;
+}
+
+message GetTokenResponse {
+    PersonalAccessToken token = 1;
+}
+
+message ListTokensRequest {
+    string user_id = 1;
+}
+
+message ListTokensResponse {
+    repeated PersonalAccessToken tokens = 1;
+}
+
+message UpdateTokenRequest {
+    bool regenerate = 1;
+    bool update_scopes = 2;
+    repeated Scope scopes = 3;
+}
+
+message UpdateTokenResponse {
+    PersonalAccessToken token = 1;
+}
+
+message DeleteTokenRequest {
+    string token_hash = 1;
+    string user_id = 2;
+}
+
+message DeleteTokenResponse {
+}
+
+message Scope {
+    repeated string options = 1;
+}
+
+message PersonalAccessToken {
```

Comment:
Each token has to have an ID

The token_hash is the secret value of the token

Reply:
You'll also need a field for the `raw` value of the token - when it's first created and we're sending it back to the user

Reply:
> We cannot use the token_hash as the ID because you can regenerate a token

You're right!